src/fu-util.c

```diff
@@ -560,6 +617,227 @@ fu_util_install_prepared (FuUtilPrivate *priv, gchar **values, GError **error)
+	// 1. copy and paste it
+	// 2. create a GetReportAsJSON dbus method
+	// 3. create a GetHistory dbus method that can marshal the FuDevice additions (etc FirmwareId) as metadata on FwupdDevice
+	// 4. move fu-pending.c to fwupd-history.c and export the API in libfwupd, creating fwupd_history_get_unreported_as_json()
+	builder = json_builder_new ();
+	json_builder_begin_object (builder);
+	json_builder_set_member_name (builder, "ReportVersion");
+	json_builder_add_int_value (builder, 1);
+	json_builder_set_member_name (builder, "MachineId");
+	json_builder_add_string_value (builder, machine_hash);
+
+	/* add each device */
+	json_builder_set_member_name (builder, "Reports");
+	json_builder_begin_array (builder);
+	for (guint i = 0; i < results->len; i++) {
+		FuDevice *dev = g_ptr_array_index (results, i);
+		if (fu_device_has_flag (dev, FWUPD_DEVICE_FLAG_REPORTED))
+			continue;
+
+		json_builder_begin_object (builder);
+
+		/* identify different devices */
+		json_builder_set_member_name (builder, "DeviceId");
+		json_builder_add_string_value (builder, fu_device_get_id (dev));
+
+		/* identify the firmware used */
+		json_builder_set_member_name (builder, "FirmwareId");
+		json_builder_add_string_value (builder, fu_device_get_firmware_id (dev));
+
+		/* set the error state of the report */
+		json_builder_set_member_name (builder, "UpdateState");
+		json_builder_add_int_value (builder, fu_device_get_update_state (dev));
+		if (fu_device_get_update_error (dev) != NULL) {
+			json_builder_set_member_name (builder, "UpdateError");
+			json_builder_add_string_value (builder, fu_device_get_update_error (dev));
+		}
+
+		/* map back to the dev type on the LVFS */
+		json_builder_set_member_name (builder, "Guid");
+		json_builder_add_string_value (builder, fu_device_get_guid_default (dev));
+
+		/* to know what plugin tried to handle the dev */
+		json_builder_set_member_name (builder, "FwupdVersion");
+		json_builder_add_string_value (builder, fu_device_get_metadata (dev, "FwupdVersion"));
+		json_builder_set_member_name (builder, "Plugin");
+		json_builder_add_string_value (builder, fu_device_get_plugin (dev));
+
+		/* report what we're trying to update *from* and *to* */
+		json_builder_set_member_name (builder, "Version");
+		json_builder_add_string_value (builder, fu_device_get_version (dev));
+		json_builder_set_member_name (builder, "VersionNew");
+		json_builder_add_string_value (builder, fu_device_get_version_new (dev));
+
+		/* to know the state of the dev we're trying to update */
+		json_builder_set_member_name (builder, "Flags");
+		json_builder_add_int_value (builder, fu_device_get_flags (dev));
+
+		/* to know when the update tried to happen, and how soon after boot */
+		json_builder_set_member_name (builder, "Created");
+		json_builder_add_int_value (builder, fu_device_get_created (dev));
+		json_builder_set_member_name (builder, "Modified");
+		json_builder_add_int_value (builder, fu_device_get_modified (dev));
+
```

Comment:
If collecting all of this level of information I think it would be good to also get device metadata.  The use case that comes to mind is the Thunderbolt/Thunderbolt-Power type of interactions.  Knowing that safe mode can be supported etc.

Reply:
I agree, but lets get the basics worked out first :)